client/consensus/beefy/src/worker.rs

```diff
@@ -872,6 +905,20 @@ where
 						return;
 					}
 				},
+				justif = gossip_proofs.next() => {
+					if let Some(justif) = justif {
+						// Gossiped justifications have already been verified by `GossipValidator`.
+						if let Err(err) = self.triage_incoming_justif(justif) {
```

Comment:
Just checking - the justification is verified by `GossipValidator` and we know it is valid. But equivocations are detected here, by `BeefyWorker`. And unless I'm missing something, they're only detected when we handle incoming votes. I wonder if it is possible for validator to make two votes and stay unnoticed because the second vote will be gossiped using finality proof? I believe that even if I'm right in my assumption, it is very unlikely scenario (given enough validators set length), so maybe something for the future

Reply:
yes, for a validator set of size `N`, it's possible now for `m` voters to "undetectably equivocate" by sending a vote for `mmr_root_2`, then including another vote for `mmr_root_1` in a proof alongside `2*N/3` other votes for `mmr_root_1`.

1. Because the precondition of required `2*N/3` "good" votes before being able to "undetectably equivocate", it means `m` can be at most `N/3` in **"perfect"** network conditions where coordination of who sees what gossipped votes is 100% controlled by dishonest validators.
2. Even in these perfect conditions where `m=N/3`, best you could do is partition the network to get `N/3` honest voting on `mmr_root_1`, another `N/3` honest voting on `mmr_root_2` and `N/3` dishonest equivocating both => at least one validator will have to _visibly_ equivocate to get over threshold.

Also, because BEEFY is based on GRANDPA, finality point.2 above is not possible, honest validators cannot be convinced to vote on anything other than what GRANDPA finalized.

Ergo, I believe this is not a problem/risk.